internal/gatewayapi/translator.go

```diff
@@ -777,6 +899,71 @@ func isValidCrossNamespaceRef(from crossNamespaceFrom, to crossNamespaceTo, refe
 	return false
 }
 
+// Checks if a hostname is valid according to RFC 1123 and gateway API's requirement that it not be an IP address
+func isValidHostname(hostname string) (valid bool, reason string) {
+	if len(hostname) == 0 || hostname == "" {
+		return false, "Hostname cannot be empty"
+	}
+
+	if len(hostname) > 255 {
+		return false, fmt.Sprintf("Hostname '%s' has a length of %d. cannot exceed 255", hostname, len(hostname))
+	}
+
+	if hostname[0] == '-' || hostname[0] == '.' {
+		return false, "Hostname cannot start with '.' or '-'"
+	}
+	if hostname[len(hostname)-1] == '-' || hostname[len(hostname)-1] == '.' {
+		return false, "Hostname cannot end with '.' or '-'"
+	}
+
+	// The only allowed characters in hostnames are a-z, A-Z, 0-9, ., and -
+	var validChar = regexp.MustCompile(`[a-zA-Z0-9.-]`).MatchString
+
+	curLabel := 0
+	for i := 0; i < len(hostname); i++ {
+		if !validChar(string(hostname[i])) {
+			return false, fmt.Sprintf("Invalid character: '%c' in hostname: '%s' at offset: %d", hostname[i], hostname, i)
+		}
+
+		if hostname[i] == '.' {
+			switch {
+			case i-curLabel > 63:
+				return false, fmt.Sprintf("Label: '%s' in hostname '%s' cannot exceed 63 characters", hostname[curLabel:i], hostname)
+			case i == curLabel:
+				return false,
+					fmt.Sprintf("Invalid character: '%c' in hostname: '%s' at offset: %d. hostname label cannot start with a '.'",
+						hostname[i],
+						hostname,
+						i,
+					)
+			case hostname[curLabel] == '-' || hostname[curLabel] == '.':
+				return false,
+					fmt.Sprintf(
+						"Label '%s' in hostname: '%s' cannot begin with a '.' or '-'",
+						hostname[curLabel:i],
+						hostname,
+					)
+
+			case hostname[i-1] == '-' || hostname[i-1] == '.':
+				return false,
+					fmt.Sprintf(
+						"Label '%s' in hostname: '%s' cannot end with a '.' or '-'",
+						hostname[curLabel:i],
+						hostname,
+					)
+			}
+			curLabel = i + 1
+		}
+	}
```

Comment:
So to sorta answer my earlier question: It looks like Kube's `IsDNS1123Subdomain` doesn't check the lengths of individual labels.  We should probably send kubenetes/kubernetes.git a PR to fix that.

But! It looks like yours doesn't quite to it right either.  If I'm reading this right, it'd allow the _last_ label to be longer than 63 chars.

Reply:
I'll fix the issue in my code and then also open a PR into kubernetes/kubetnetes.git, but for now I'll just leave a TODO comment saying that we should switch to `IsDNS1123Subdomain` once that submission lands.

Reply:
another option to consider https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/pkg/config/labels/instance.go#L95

Reply:
We don't need the wildcard check since that's invalid for a redirect, but the logic is now a combination of `IsDNS1123Subdomain` and a check on the individual label length.